### ISSUE TITLE
Function argument with object rest spread generates wrong type error

### ISSUE THREAD

**Issue by zamb3zi** [Microsoft/TypeScript#25217]
**TypeScript Version:**  2.9.1-insiders.20180525

**Search Terms:** argument rest spread object

**Code**

```ts
const f = (options: { foo?: number }) => { };
let o: {
    foo?: number;
    bar?: string
} = {};
f(o); // ok
f({ ...o }); // error: bar does not exist in type { foo?: number }.
f({ ...o, foo: 5 }); // same error. this is the use case.
let p = { ...o };
f(p); // ok
f(p = { ...o, foo: 5 }); // also ok
```

**Expected behavior:**

`{...o}` has the same type as `o` and should be interchangeable as function argument.

**Actual behavior:**

Incorrect error is generated.

**Playground Link:** 

[https://www.typescriptlang.org/play/index.html#src=const%20f%20%3D%20(options%3A%20%7B%20foo%3F%3A%20number%20%7D)%20%3D%3E%20%7B%20%7D%3B%0D%0Alet%20o%3A%20%7B%0D%0A%20%20%20%20foo%3F%3A%20number%3B%0D%0A%20%20%20%20bar%3F%3A%20string%0D%0A%7D%20%3D%20%7B%7D%3B%0D%0Af(o)%3B%20%2F%2F%20ok%0D%0Af(%7B%20...o%20%7D)%3B%20%2F%2F%20error%3A%20bar%20does%20not%20exist%20in%20type%20%7B%20foo%3F%3A%20number%20%7D.%0D%0Af(%7B%20...o%2C%20foo%3A%205%20%7D)%3B%20%2F%2F%20same%20error.%20this%20is%20the%20use%20case.%0D%0Alet%20p%20%3D%20%7B%20...o%20%7D%3B%0D%0Af(p)%3B%20%2F%2F%20ok%0D%0Af(p%20%3D%20%7B%20...o%2C%20foo%3A%205%20%7D)%3B%20%2F%2F%20also%20ok](url)

**Related Issues:**

**Comment by RyanCavanaugh** [comment]
@sandersn please apply the appropriate label to this issue :

**Comment by sandersn** [comment]
Duplicate of #19775

Probably won't be fixed until we have exact types, because people who don't want excess property errors currently have two workarounds (a [deprecated] compiler flag and extract-local), but people who do want excess property errors would have no way to cause them if this bug is fixed.

**Comment by zamb3zi** [comment]
Thanks a lot and sorry I couldn't find the duplicate. For anyone else landing here, some workarounds are:
```ts
const f = <O extends { foo?: number }>(options: O) => { };
```
OR
```ts
const f = (options: { foo?: number }) => { };
const ignoreExcess = <T>(arg: T) => arg;
let o: {
    foo?: number;
    bar?: string
} = {};
f(ignoreExcess({ ...o, foo: 5 })); // ok
```

**Comment by sandersn** [comment]
You can also do

```ts
const f = (options: { foo?: number }) => { };
let o: {
    foo?: number;
    bar?: string
} = {};
const ignoreExcess = { ...o, foo: 5 };
f(ignoreExcess); // ok
```

Which is probably cheaper at runtime, depending on how much the VM tries to optimise.
